modules/consensus/setdb.go

```diff
@@ -0,0 +1,226 @@
+// future
+func (cs *ConsensusSet) pbToBn(pb *processedBlock) blockNode {
+	parent, exists := cs.blockMap[pb.Parent]
+	if !exists {
+		panic("block parent not in consensus set")
+	}
+
+	bn := blockNode{
+		block:  pb.Block,
+		parent: parent,
+
+		height:      pb.Height,
+		depth:       pb.Depth,
+		childTarget: pb.ChildTarget,
+
+		diffsGenerated:            pb.DiffsGenerated,
+		siacoinOutputDiffs:        pb.SiacoinOutputDiffs,
+		fileContractDiffs:         pb.FileContractDiffs,
+		siafundOutputDiffs:        pb.SiafundOutputDiffs,
+		delayedSiacoinOutputDiffs: pb.DelayedSiacoinOutputDiffs,
+		siafundPoolDiffs:          pb.SiafundPoolDiffs,
+
+		consensusSetHash: pb.ConsensusSetHash,
+	}
+	return bn
+}
+
+func openDB(filename string) (*setDB, error) {
+	db, err := persist.OpenDatabase(meta, filename)
+	if err != nil {
+		return nil, err
+	}
+
+	var buckets []string = []string{
+		"Path", "BlockMap",
+	}
+
+	// Create buckets
+	err = db.Update(func(tx *bolt.Tx) error {
+		for _, bucketName := range buckets {
+			_, err := tx.CreateBucketIfNotExists([]byte(bucketName))
+			if err != nil {
+				return err
+			}
+		}
+		return nil
+	})
+	return &setDB{db}, nil
+}
+
+// addItem and getItem are part of consensus due to stricter error conditions.
+//
+// addItem should only be called from this file, and adds a new item
+// to the database
+func (db *setDB) addItem(bucket string, key, value interface{}) error {
+	v := encoding.Marshal(value)
+	k := encoding.Marshal(key)
+	return db.Update(func(tx *bolt.Tx) error {
+		b := tx.Bucket([]byte(bucket))
+		if build.DEBUG {
+			if b == nil {
+				panic(errNilBucket)
+			}
+			i := b.Get(k)
+			if i != nil {
+				panic(errBadSetInsert)
+			}
+		}
+		return b.Put(k, v)
+	})
+}
+
+func (db *setDB) getItem(bucket string, key interface{}) (item []byte, err error) {
+	k := encoding.Marshal(key)
+	err = db.View(func(tx *bolt.Tx) error {
+		b := tx.Bucket([]byte(bucket))
+		if build.DEBUG {
+			if b == nil {
+				panic(errNilBucket)
+			}
+		}
+		item = b.Get(k)
+		if build.DEBUG {
+			if item == nil {
+				panic(errNilItem)
+			}
+		}
+		return nil
+	})
+	return
+}
+
+// AddBlock inserts a block into the database at the "end" of the chain, i.e.
+// the current height + 1.
+func (db *setDB) addPath(block types.Block) error {
+	value := encoding.Marshal(block.ID())
+	return db.Update(func(tx *bolt.Tx) error {
+		b := tx.Bucket([]byte("Path"))
+		key := encoding.EncUint64(uint64(b.Stats().KeyN))
+		return b.Put(key, value)
+	})
+}
+
+// RemoveBlock removes a block from the "end" of the chain, i.e. the block
+// with the largest height.
+func (db *setDB) rmPath() error {
+	return db.Update(func(tx *bolt.Tx) error {
+		b := tx.Bucket([]byte("Path"))
+		key := encoding.EncUint64(uint64(b.Stats().KeyN - 1))
+		return b.Delete(key)
+	})
+}
+
+// path retreives the block id of a block at a given hegiht from the path
+func (db *setDB) path(h types.BlockHeight) (id types.BlockID, err error) {
```

Comment:
I don't think that this function name is appropriate.

perhaps blockAtHeight, which is also a bit awkward but at least it's more descriptive.